Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -65,30 +70,56 @@ namespace internal {
 // An operation that notifies the corresponding stream on its completion (via
 // `GrpcStreamDelegate`). The stream is guaranteed to be valid as long as the
 // operation exists.
-class StreamOperation : public GrpcOperation {
+class StreamOperation : public GrpcOperation,
+                        public std::enable_shared_from_this<StreamOperation> {
  public:
   StreamOperation(GrpcStreamDelegate&& delegate,
                   grpc::GenericClientAsyncReaderWriter* call,
+                  AsyncQueue* firestore_queue,
                   GrpcCompletionQueue* grpc_queue)
-      : delegate_{std::move(delegate)}, call_{call}, grpc_queue_{grpc_queue} {
+      : delegate_{std::move(delegate)},
+        call_{call},
+        firestore_queue_{firestore_queue},
+        grpc_queue_{grpc_queue} {
+  }
+
+  void Cancel() override {
+    firestore_queue_->VerifyIsCurrentQueue();
+    delegate_.RemoveOperation(this);
   }
 
   void Execute() override {
+    firestore_queue_->VerifyIsCurrentQueue();
+
+    // TODO OBC: grpc_queue should no longer be necessary
     if (!grpc_queue_->is_shut_down()) {
       DoExecute(call_);
     }
   }
 
-  void Complete(bool ok) override {
-    if (ok) {
-      DoComplete(&delegate_);
-    } else {
-      // Failed operation means this stream is unrecoverably broken; use the
-      // same error-handling policy for all operations.
-      delegate_.OnOperationFailed();
-    }
+  std::future<void> get_future() {
+    return completed_.get_future();
+  }
 
-    delegate_.RemoveOperation(this);
+  void Complete(bool ok) override {
+    std::weak_ptr<StreamOperation> weak_self{shared_from_this()};
```

Comment:
@wilhuff Gil, so I tried out the change we discussed. There was an unforeseen complication, however: when we're calling `GrpcStream::Finish` on Firestore queue, there might already be an operation pulled out from GRPC completion queue and scheduled for later execution on Firestore queue (imagine the state of the queue is something like this:)
```
{ [GrpcStream::Finish()], [Foo], [Bar], [StreamRead::Complete()], ... }
    ^
```
This means that if the stream is destroyed immediately after `GrpcStream::Finish`, the stream will also destroy its operations, so when execution goes to `StreamRead::Finish`, it will refer to deleted memory.

A straightforward way around this is using `shared_ptr`/`weak_ptr` to manage operations' lifetime, which is what I implemented here: `GrpcStream` stores (and releases) `shared_ptr`s to operations, and operations enqueue `weak_ptr`s to themselves on the Firestore queue and then check if `weak_ptr` still refers to a valid instance -- sort of like the Objective-C pattern.

The downside is in additional complications: there's now a future/promise mechanism plus shared pointers. I guess the upside is that ownership is more local to `GrpcStream`. Let me know what you think.